pkg/trustedresources/verifier/verifier_test.go

```diff
@@ -0,0 +1,419 @@
+				t.Errorf("FromPolicy got empty verifiers")
+			}
+			if err != nil {
+				t.Errorf("FromPolicy got: %v, want: nil", err)
+			}
+		})
+	}
+}
+
+func TestFromPolicy_Error(t *testing.T) {
+	tcs := []struct {
+		name          string
+		policy        *v1alpha1.VerificationPolicy
+		expectedError error
+	}{{
+		name: "hash algorithm is invalid",
+		policy: &v1alpha1.VerificationPolicy{
+			Spec: v1alpha1.VerificationPolicySpec{
+				Authorities: []v1alpha1.Authority{
+					{
+						Key: &v1alpha1.KeyRef{
+							Data:          "inlinekey",
+							HashAlgorithm: "sha1",
+						},
+					},
+				},
+			},
+		},
+		expectedError: errorAlgorithmInvalid,
+	}, {
+		name: "key is empty",
+		policy: &v1alpha1.VerificationPolicy{
+			Spec: v1alpha1.VerificationPolicySpec{
+				Authorities: []v1alpha1.Authority{
+					{
+						Key: &v1alpha1.KeyRef{},
+					},
+				},
+			},
+		},
+		expectedError: errorEmptyKey,
+	}, {
+		name: "authority is empty",
+		policy: &v1alpha1.VerificationPolicy{
+			Spec: v1alpha1.VerificationPolicySpec{
+				Authorities: []v1alpha1.Authority{},
+			},
+		},
+		expectedError: errorEmptyPublicKeys,
+	}, {
+		name: "from data error",
+		policy: &v1alpha1.VerificationPolicy{
+			Spec: v1alpha1.VerificationPolicySpec{
+				Authorities: []v1alpha1.Authority{
+					{
+						Key: &v1alpha1.KeyRef{
+							Data:          "inlinekey",
+							HashAlgorithm: "sha256",
+						},
+					},
+				},
+			},
+		},
+		expectedError: ErrorDecodeKey,
+	}, {
+		name: "from secret error",
+		policy: &v1alpha1.VerificationPolicy{
+			Spec: v1alpha1.VerificationPolicySpec{
+				Authorities: []v1alpha1.Authority{
+					{
+						Key: &v1alpha1.KeyRef{
+							SecretRef: &v1.SecretReference{
+								Name:      "wrongSecret",
+								Namespace: "wrongNamespace",
+							},
+						},
+					},
+				},
+			},
+		},
+		expectedError: errorSecretNotFound,
+	}}
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			_, err := FromPolicy(context.Background(), fakek8s.NewSimpleClientset(), tc.policy)
+			if !errors.Is(err, tc.expectedError) {
+				t.Errorf("FromPolicy got: %v, want: %v", err, tc.expectedError)
+			}
+		})
+	}
+}
+
+func TestFromKeyRef_Success(t *testing.T) {
+	ctx := context.Background()
+	_, keypath := test.GetSignerFromFile(ctx, t)
+
+	_, pub, err := test.GenerateKeys(pass("1234"))
+	if err != nil {
+		t.Errorf("failed to generate keys: %v", err)
+	}
+	secretData := &v1.Secret{
+		Data: map[string][]byte{
+			"data": pub,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "secret",
+			Namespace: namespace}}
+
+	k8sclient := fakek8s.NewSimpleClientset(secretData)
+
+	tcs := []struct {
+		name   string
+		keyref string
+	}{{
+		name:   "key in file",
+		keyref: keypath,
+	}, {
+		name:   "key in secret",
+		keyref: fmt.Sprintf("k8s://%s/secret", namespace),
+	}}
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			verifier, err := fromKeyRef(ctx, tc.keyref, crypto.SHA256, k8sclient)
```

Comment:
same question re. can we check anything on the verifier itself - ill stop repeating that feedback but i think it would apply to other tests below as well

Reply:
done